Implements better (maybe) lag compensation for gliding

LemonInTheDark opened the pull request:
Our existing lag compensation is utter dogwater, it works off a 10 second average that was taken between 0 and 10 seconds ago. Overshooting leads to stutter steps, which are highly unpleasent, and easy to achive since any spike in overtime will lead to an overshoot.

What I'm doing here is injesting cpu_usage, deaveraging it using a world/Tick override (it has an 16 frame moving avg), then taking an average of that and dropping noisy bits.

I am not the best at stats so this implementation might kinda suck, but it's better then what we have now.

I'm also trying out spike correction, IE accounting for the delay a lagspike causes.

This is less of an issue then jumping is because you just end up staying on a turf for a few ticks longer, but if we cna just slow people down some I think it's worthwhile.

I am less confident in this concept/implementation then I am with the tick correction, but we'll see.

I'm putting this up to test on live/motion sick people, should not be merged.

LemonInTheDark committed:
Implements better (maybe) lag compensation for gliding

Our existing lag compensation is utter dogwater, it works off a 10
second average that was taken between 0 and 10 seconds ago. Overshooting
leads to stutter steps, which are highly unpleasent, and easy to achive
since any spike in overtime will lead to an overshoot.

What I'm doing here is injesting cpu_usage, deaveraging it using a
world/Tick override (it has an 16 frame moving avg), then taking an
average of that and dropping noisy bits.

I am not the best at stats so this implementation might kinda suck, but
it's better then what we have now.

I'm also trying out spike correction, IE accounting for the delay a
lagspike causes.

This is less of an issue then jumping is because you just end up staying
on a turf for a few ticks longer, but if we cna just slow people down
some I think it's worthwhile.

I am less confident in this concept/implementation then I am with the
tick correction, but we'll see.

I'm putting this up to test on live/motion sick people, should not be
merged.

ZephyrTFA commented:
I'm not sure if this is a TM only PR but it has testing code in it still so just as a marker I'm putting DNM and TM on it

LemonInTheDark commented:
it is for now at least

LemonInTheDark committed:
what the fuck

LemonInTheDark committed:
size increase

LemonInTheDark committed:
Changes avg method to keep a fix sized window

I hate byond

LemonInTheDark committed:
whoop

SyncIt21 commented on code/controllers/subsystem/time_track.dm:
```diff
@@ -93,7 +93,6 @@ SUBSYSTEM_DEF(time_track)
 		time_dilation_avg_fast = MC_AVERAGE_FAST(time_dilation_avg_fast, time_dilation_current)
 		time_dilation_avg = MC_AVERAGE(time_dilation_avg, time_dilation_avg_fast)
 		time_dilation_avg_slow = MC_AVERAGE_SLOW(time_dilation_avg_slow, time_dilation_avg)
-		GLOB.glide_size_multiplier = (current_byondtime - last_tick_byond_time) / (current_realtime - last_tick_realtime)
```
Very interesting. Did you know this line can runtime? I reported this as an issue #75221

LemonInTheDark replied:
I uh, don't know how that's possible, but I suppose so yes

LemonInTheDark committed:
IT WORKED!!!

LemonInTheDark committed:
oh whoops\

LemonInTheDark committed:
Lemon's amazing magic mob mover

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into dynamic-gliding

LemonInTheDark committed:
Implements potentially better correction and debugging

Glide corrects as if we are a tick in the past (since our info is a tick
out of date), draws more debug info (and draws it to a hidden plane
master) and locks active correction behind a global value

LemonInTheDark commented:
updated to include maybe better tick correction and more debug control, I intend to tm this when next convenient and do some pokin around

LemonInTheDark committed:
Better debug + glide text

LemonInTheDark committed:
Fixes people seeing my shit, better control, more complete tick readout to shittest

LemonInTheDark committed:
fuck

LemonInTheDark committed:
Noise control, verb

Floors real cpu values with the tick usage/map cpu from the last tick
This serves as a worst case minimum and should bound any error
introduced by floating point bullshit to a particular bound (since error
compounds here)

Adds a verb that toggles on/off the cpu display, so players are not
subjected to it for no reason

LemonInTheDark committed:
Reserves more time for verbs, adds more tracking, autocorrection mode

Verbs take an undefined amount of time (we should really track/manage
this better) and if we're gonna be consuming 100% of the tick we should
limit that (more on that later)

We have map cpu, tick cost, and deaveraged cpu, so we can work out
ROUGHLY how much time we spend on verbs + other bullshit. Should display
that

Adds a new global that enables cpu stabailization, queuing up
/world/Tick() to consume a constant amount of cpu (since that stabalizes
the deltas between sendmaps calls, reducing signficantly visible lag)

LemonInTheDark commented:
has done its job, debug code may be useful in future who knows
